com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs

```diff
@@ -536,73 +553,112 @@ protected void ChannelsToTexture(int channelIndex, int numChannelsToAdd)
         /// </summary>
         /// <param name="foundColliders">Array of the colliders found within the cell</param>
+        /// <param name="numFound">Number of colliders found.</param>
         /// <param name="cellIndex">The index of the cell</param>
         /// <param name="cellCenter">The center position of the cell</param>
-        protected virtual void ParseColliders(Collider[] foundColliders, int cellIndex, Vector3 cellCenter)
+        protected virtual void ParseColliders(Collider[] foundColliders, int numFound, int cellIndex, Vector3 cellCenter)
         {
-            GameObject currentColliderGo = null;
+            Profiler.BeginSample("GridSensor.ParseColliders");
             GameObject closestColliderGo = null;
-            Vector3 closestColliderPoint = Vector3.zero;
-            float distance = float.MaxValue;
-            float currentDistance = 0f;
+            var distance = float.MaxValue;
 
-            for (int i = 0; i < foundColliders.Length; i++)
+            for (var i = 0; i < numFound; i++)
             {
-                currentColliderGo = foundColliders[i].gameObject;
+                var currentColliderGo = foundColliders[i].gameObject;
 
                 // Continue if the current collider go is the root reference
-                if (currentColliderGo == rootReference)
+                if (ReferenceEquals(currentColliderGo, rootReference))
                     continue;
 
-                closestColliderPoint = foundColliders[i].ClosestPointOnBounds(cellCenter);
-                currentDistance = Vector3.Distance(closestColliderPoint, rootReference.transform.position);
+                var closestColliderPoint = foundColliders[i].ClosestPointOnBounds(cellCenter);
+                var currentDistance = (closestColliderPoint - rootReference.transform.position).sqrMagnitude;
 
                 // Checks if our colliders contain a detectable object
-                if ((Array.IndexOf(DetectableObjects, currentColliderGo.tag) > -1) && (currentDistance < distance))
+                var index = -1;
+                for (var ii = 0; ii < DetectableObjects.Length; ii++)
+                {
+                    if (currentColliderGo.CompareTag(DetectableObjects[ii]))
+                    {
+                        index = ii;
+                        break;
+                    }
+                }
+                if (index > -1 && currentDistance < distance)
                 {
                     distance = currentDistance;
                     closestColliderGo = currentColliderGo;
                 }
             }
 
-            if (closestColliderGo != null)
-                LoadObjectData(closestColliderGo, cellIndex, distance / SphereRadius);
+            if (!ReferenceEquals(closestColliderGo, null))
+                LoadObjectData(closestColliderGo, cellIndex, (float)Math.Sqrt(distance) / SphereRadius);
```

Comment:
microoptimization: store `1.0f / SphereRadius`  as `InverseSphereRadius`, and multiply by that everywhere instead.